rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,674 @@
+addition, the specified behavior can subsequently be used as an abstraction of
+the function's behavior at call sites.
+
+The lifecycle of a contract is split roughly into three phases: specification,
+verification and stubbing. Which we will explore on this simple example:
+
+```rs
+fn my_div(dividend: u32, divisor: u32) -> u32 {
+  dividend / divisor
+}
+```
+
+1. In the first phase we **specify** the approximation. Kani provides two new
+   annotations: `requires` (preconditions) to describe the expectations this
+   function has as to the calling context and `ensures` (postconditions) which
+   approximates function outputs in terms of function inputs.
+
+   ```rs
+   #[kani::requires(divisor != 0)]
+   #[kani::ensures(result <= dividend)]
+   fn my_div(dividend: u32, divisor: u32) -> u32 {
+     dividend / divisor
+   }
+   ```
+  
+   `requires` here indicates this function expects its `divisor` input to never
+   be 0, or it will not execute correctly (i.e. panic).
+
+   `ensures` puts a bound on the output, relative to the `dividend` input.
+
+   Conditions in contracts are plain Rust expressions which can reference the
+   function arguments and, in case of `ensures`, the result of the function as a
+   special `result` variable (see [open questions](#open-questions) on a
+   discussion about (re)naming). Syntactically Kani supports any Rust
+   expression, including function calls, defining types etc. However they must
+   be side-effect free[^side-effects] or Kani will throw a compile error.
+
+   Multiple `requires` and `ensures` clauses are allowed on the same function,
+   they are implicitly logically conjoined.
+
+
+2. Next, Kani must make sure that the approximation we specified actually holds
+   by **checking** the contract against the implementation. This is in contrast
+   to ["stubbing"][stubbing], where the approximation is blindly trusted.
+
+   The contract must always overapproximate the function to guarantee soundness.
+   To facilitate the check Kani needs a suitable environment to verify the
+   function in. For this proposal the environment must be provided by us (the
+   users). See [future possibilities](#future-possibilities) for a discussion
+   about the arising soundness issues and their remedies.
+
+   We provide the checking environment for our contract with a special new
+   `proof_for_contract` harness.
+
+   ```rs
+   #[kani::proof_for_contract(my_div)]
+   fn my_div_harness() {
+     my_div(kani::any(), kani::any())
+   }
+   ```
+
+   Similar to a verification harness for any other function, we are supposed to
+   create all possible input combinations the function can encounter, then call
+   the function at least once with those abstract inputs. If we forget to call
+   `my_div` Kani reports an error.
+   
+   Unlike a unit-test we can however elide any checks of the output and
+   post-call state. Instead Kani uses the conditions we specified in the
```

Comment:
Maybe just say what happens when checking a contract without explaining it as a diff to what happens with traditional harnesses